### DISCUSSION TITLE
Detect whether an optional struct field is present but non-concrete

### DISCUSSION THREAD

**Discussion by seh** [cue-lang/cue#2097] [Q&A]
Should it be possible to distinguish the difference between a struct’s optional field being absent and the optional field being present as a non-concrete constraint? [In the CUE Playground](https://tip.cuelang.org/play/?id=YnOFL9ModT7#cue@export@cue), we can see that regardless of whether the struct field is optional, checking whether a non-concrete constraint value is bottom (`_|_`) yields true.

Is there some way to distinguish between the optional field being wholly absent and there being a present-but-non-concrete constraint? That is, in the following example, how can we detect that the `present` struct value constrains field "b?"
```cue
#T: {
  a: string
  b?: string
}

absent: #T & {
  a: "abc"
}

present: #T & {
  a: "abc"
  b: !="def"
}

concrete: #T & {
  a: "abc"
  b: !="def"
  b: "efg"
}

detectAbsent: absent.b == _|_
// Why is this value false?
detectPresent: present.b != _|_
detectConcrete: concrete.b != _|_
```
#943 promises something like an `exists` built-in function, but states that for now we can compare against `_|_`. I'm finding that comparing against `_|_` isn't good enough for this case.

**Comment by myitcv** [comment]
Hi @seh!

> Should it be possible to distinguish the difference between a struct’s optional field being absent and the optional field being present as a non-concrete constraint?

Yes. That's the subject of the proposed builtins in #943. Specifically, `isconcrete()` and `exists()` (which might be called `isdefined()` following a more precise definition of defined fields in https://github.com/cue-lang/cue/discussions/1951). In the context of your example:

```
#T: {
  a: string
  b?: string
}

absent: #T & {
  a: "abc"
}

present: #T & {
  a: "abc"
  b: !="def"
}

concrete: #T & {
  a: "abc"
  b: !="def"
  b: "efg"
}

exists_absent_b: exists(absent.b) == false
exists_present_b: exists(present.b) == true
concrete_present_b: isconcrete(present.b) == false
concrete_concrete_b: isconcrete(concrete.b) == true
```
